From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Fri, 28 May 2021 21:07:08 -0400
Subject: [PATCH] Fox API


diff --git a/src/main/java/net/minecraft/world/entity/animal/EntityFox.java b/src/main/java/net/minecraft/world/entity/animal/EntityFox.java
index b7fa24318ef43918b6b10ff4ea8acb960527296e..de142a7387740bfd9c656b8cd46a08554fc3ce16 100644
--- a/src/main/java/net/minecraft/world/entity/animal/EntityFox.java
+++ b/src/main/java/net/minecraft/world/entity/animal/EntityFox.java
@@ -397,18 +397,22 @@ public class EntityFox extends EntityAnimal {
         this.d(1, flag);
     }
 
+    public boolean isFaceplanted() {return eN();} // Paper - OBFHELPER
     public boolean eN() {
         return this.t(64);
     }
 
+    public void setFaceplanted(boolean faceplanted) {x(faceplanted);} // Paper - OBFHELPER
     private void x(boolean flag) {
         this.d(64, flag);
     }
 
+    public boolean isDefending() {return fb();} // Paper - OBFHELPER
     private boolean fb() {
         return this.t(128);
     }
 
+    public void setDefending(boolean defending) {y(defending);} // Paper - OBFHELPER
     private void y(boolean flag) {
         this.d(128, flag);
     }
@@ -541,10 +545,12 @@ public class EntityFox extends EntityAnimal {
         ((EntityFox) entityinsentient).b(entityhuman.getUniqueID());
     }
 
+    public boolean isLeaping() {return eO();} // Paper - OBFHELPER
     public boolean eO() {
         return this.t(16);
     }
 
+    public void setLeaping(boolean leaping) {u(leaping);} // Paper - OBFHELPER
     public void u(boolean flag) {
         this.d(16, flag);
     }
@@ -561,10 +567,12 @@ public class EntityFox extends EntityAnimal {
         return this.t(4);
     }
 
+    public void setInterested(boolean interested) {w(interested);} // Paper - OBFHELPER
     public void w(boolean flag) {
         this.d(8, flag);
     }
 
+    public boolean isInterested() {return eW();} // Paper - OBFHELPER
     public boolean eW() {
         return this.t(8);
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftFox.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftFox.java
index 362ac8e43096995184c2d13ccdd69a84103a791c..2511d7009e7d6c3459acb899274c1a715545e7e2 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftFox.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftFox.java
@@ -115,4 +115,45 @@ public class CraftFox extends CraftAnimals implements Fox {
 
         getHandle().getDataWatcher().set(EntityFox.SECOND_TRUSTED_PLAYER, player == null ? Optional.empty() : Optional.of(player.getUniqueId()));
     }
+    // Paper start - Add more fox behavior API
+    @Override
+    public void setInterested(boolean interested) {
+        getHandle().setInterested(interested);
+    }
+
+    @Override
+    public boolean isInterested() {
+        return getHandle().isInterested();
+    }
+
+    @Override
+    public void setLeaping(boolean leaping) {
+        getHandle().setLeaping(leaping);
+    }
+
+    @Override
+    public boolean isLeaping() {
+        return getHandle().isLeaping();
+    }
+
+    @Override
+    public void setDefending(boolean defending) {
+        getHandle().setDefending(defending);
+    }
+
+    @Override
+    public boolean isDefending() {
+        return getHandle().isDefending();
+    }
+
+    @Override
+    public void setFaceplanted(boolean faceplanted) {
+        getHandle().setFaceplanted(faceplanted);
+    }
+
+    @Override
+    public boolean isFaceplanted() {
+        return getHandle().isFaceplanted();
+    }
+    // Paper end - Add more fox behavior API
 }
